src/components/date-picker/calendar.tsx

```diff
@@ -240,41 +286,79 @@ export class Calendar extends React.Component<CalendarProps, CalendarState> {
         }
     }
 
-    private toggleMonthView = () => {
-        this.setState({showMonthView: !this.state.showMonthView});
+    private toggleDateSelectView = () => {
+        if (this.calendarState.view === 'year') {
+            this.calendarState.view = 'default';
+            this.resetViewDate();
+        } else if (this.calendarState.view === 'default') {
+            this.calendarState.view = 'month';
+        } else if (this.calendarState.view === 'month') {
+            this.calendarState.view = 'year';
+        }
     }
 
-    private onSelectMonth: React.EventHandler<React.SyntheticEvent<Element>> = event => {
-        event.preventDefault();
-        this.toggleMonthView();
+    private closeDateSelectView = () => {
+        if (this.calendarState.view === 'year') {
+            this.calendarState.view = 'month';
+        } else if (this.calendarState.view === 'month') {
+            this.calendarState.view = 'default';
+        }
+    }
 
-        const date = new Date(this.props.value.getFullYear(),
+    private resetViewDate = () => {
+        this.calendarState.viewDate = this.props.value;
+    }
+
+    private onSelectMonth: React.EventHandler<React.SyntheticEvent<Element>> = event => {
+        const date = new Date(this.calendarState.viewDate.getFullYear(),
             monthNames.indexOf((event.target as HTMLSpanElement).textContent!),
-            this.props.value.getDate());
+            this.calendarState.viewDate.getDate());
+
+        this.props.updateDropdownDate(date);
+        this.calendarState.viewDate = date;
+        this.closeDateSelectView();
+    }
+
+    private onSelectYear: React.EventHandler<React.SyntheticEvent<Element>> = event => {
+        const date = new Date(parseInt((event.target as HTMLSpanElement).textContent!, 10),
+            this.calendarState.viewDate.getMonth(),
+            this.calendarState.viewDate.getDate());
 
         this.props.updateDropdownDate(date);
+        this.calendarState.viewDate = date;
+        this.closeDateSelectView();
     }
 
     private goToNextMonth: React.EventHandler<React.SyntheticEvent<Element>> = event => {
         event.preventDefault();
-        const nextDate: Date = this.state.showMonthView
-            ? new Date(this.props.value.getFullYear() + 1, this.props.value.getMonth(), 1)
-            : getMonthFromOffset(new Date(this.props.value.getFullYear(), this.props.value.getMonth(), 1), 1);
 
-        this.props.updateDropdownDate(nextDate);
+        if (this.calendarState.view !== 'default') {
```

Comment:
Would be nice to flip the condition `if (view === default) { do stuff } else { do other stuff }`